src/mygrad/typing.py

```diff
@@ -0,0 +1,6 @@
+from typing import Union, Sequence, Any
+from numbers import Real
+from numpy import ndarray
+
+
+ArrayLike = Union[Real, ndarray, Sequence[Any]]
```

Comment:
```python
if typing.TYPE_CHECKING:
    from mygrad.tensor_base import Tensor

ArrayLike = Union[Real, ndarray, "Tensor", Sequence[Any]]
```

is inelegant, but would simplify *many* other type-annotations.  The import cycle only breaks things if it's executed at runtime, after all!

```python
# Sequence[Any] is imprecise; we actually only want equally-nested sequences that end in Real
s1 = Sequence[Real]
s2 = Sequence[s1]
...
s32 = Sequence[s31]
seq_arraylike = Union[s1, s2, ..., s32]  # Sequence[Union[s1, s2]] is *not* valid!

ArrayLike = Union[Real, ndarray, seq_arraylike]
```

is considerably more verbose still, but IMO not unreasonable: it's nicely isolated in your typing helpers module, and can catch a range of issues with element types and ragged-dimensional arrays.